### PR TITLE
GH-26: Add SMS API

### PR DESCRIPTION
**Pull request by mattleibow**
### Description of Change ###

Adding the SMS API as designed in #26.

### Bugs Fixed ###

- Related to issue #26

Provide links to issues here. Ensure that a GitHub issue was created for your feature or bug fix before sending PR.

### API Changes ###

Adding the SMS API as designed in #26.

### Behavioral Changes ###

Describe any non-bug related behavioral changes that may change how users app behaves when upgrading to this version of the codebase.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [X] Has samples (if omitted, state reason in description)
- [X] Rebased on top of master at time of PR
- [X] Changes adhere to coding standard
- [x] Updated documentation ([see walkthrough](https://github.com/xamarin/Caboodle/wiki/Documenting-your-code-with-mdoc))

### CODE REVIEW

File: Caboodle/Sms/Sms.netstandard.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Threading.Tasks;
+
+namespace Microsoft.Caboodle
+{
+    public static partial class Sms
+    {
+        public static bool IsComposeSupported
```
Comment (Redth) [inline Caboodle/Sms/Sms.netstandard.cs]: I think we still need to decide if a bool and this naming is what we want to end up with.

Comment (mattleibow) [reply Caboodle/Sms/Sms.netstandard.cs]: Yeah. We may want to have some support level thing: Unsupported, Supported, SupportedWithBackgroundSend or something along those lines. This is more a feature that works when there is a SIM card of some sort.

Comment (jamesmontemagno) [reply Caboodle/Sms/Sms.netstandard.cs]: I think that probably each class should be small enough that a single Enum should represent what it needs...

So we could introduce "SmsBackground" for instance and that is a new API with a new Enum. I think that is almost better to be honest with you.

File: Caboodle/Sms/Sms.ios.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            var controller = Platform.GetCurrentViewController();
+
+            // show the controller
+            var tcs = new TaskCompletionSource<bool>();
+            var messageController = new MFMessageComposeViewController
+            {
+                Body = message.Body,
+                Recipients = new[] { message.Recipient }
+            };
+            messageController.Finished += (sender, e) =>
```
Comment (jamesmontemagno) [inline Caboodle/Sms/Sms.ios.cs]: Does this get triggered no matter what? even if it is canceled? Is it better ot fire and forget?

File: Caboodle/Sms/Sms.shared.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    {
+    }
+
+    public class SmsMessage
+    {
+        public SmsMessage()
+        {
+        }
+
+        public SmsMessage(string body, string recipient)
```
Comment (jamesmontemagno) [inline Caboodle/Sms/Sms.shared.cs]: Should there be one with just the body? also what is a recipient? phone number?

File: Caboodle/Sms/Sms.shared.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        public string Body { get; set; }
+
+        public string Recipient { get; set; }
+
+        internal void Validate()
+        {
+            if (string.IsNullOrWhiteSpace(Body))
+                throw new ArgumentException("SMS body must not be empty.", nameof(Body));
+
+            if (string.IsNullOrWhiteSpace(Recipient))
```
Comment (jamesmontemagno) [inline Caboodle/Sms/Sms.shared.cs]: I think it can be empty. I think that is fine.

File: DeviceTests/Caboodle.DeviceTests.Shared/Utils.cs [hunk 5]
```diff
@@ -1,10 +1,14 @@
 ﻿using System;
 using System.Threading.Tasks;
+using Microsoft.Caboodle;
 
 namespace Caboodle.DeviceTests
 {
     public class Utils
     {
+        public static bool IsiOSSimulator
```
Comment (jamesmontemagno) [inline DeviceTests/Caboodle.DeviceTests.Shared/Utils.cs]: Could this be a method that we pass in the device platform?

File: Caboodle/Sms/Sms.ios.cs (hunk 2 above)
Comment (mattleibow) [reply Caboodle/Sms/Sms.ios.cs]: This appears to be the case. The docs do say we have to explicitly hide the controller in this event and we are provided a result (success, canceled, failed)

File: Caboodle/Sms/Sms.shared.cs (hunk 3 above)
Comment (mattleibow) [reply Caboodle/Sms/Sms.shared.cs]: we could have overloads for each combo, or we could just let the devs use the properties.
The recipient... each platform just takes a string, so we could leave it as is and hope for the best :)

File: DeviceTests/Caboodle.DeviceTests.Shared/Utils.cs (hunk 5 above)
Comment (mattleibow) [reply DeviceTests/Caboodle.DeviceTests.Shared/Utils.cs]: This was more just for the tests, as the iOS sim was quite different

Comment (jamesmontemagno) [reply DeviceTests/Caboodle.DeviceTests.Shared/Utils.cs]: ahhh gotcha <3 I see it is devicetests cool.

**Comment by jamesmontemagno** [comment]
My only feedback here is to allow empty recipients in general. Also... does it hurt if the message is blank? Maybe they just want to launch the sms app?

**Comment by Redth** [comment]
1. Let's have a `ComposeAsync()` (but please double check we can open sms compose on all platforms with no recipient/body.
2. Don't require recipient/body.
3. Make `IsComposeSupported` internal for now, and check it when we execute `ComposeAsync` and throw the same exception across platforms if it's unsupported at runtime.

**Comment by mattleibow** [comment]
I think this is done.
I haven't really been able to properly test the Android part - I think my old Android device is a bit messed - I don't appear to have a messaging app... Or a SIM card for it. I could reset and use my SIM, but there are others that are Android users, I think? Is it possible for someone with a real device to test?

**Pull request merged**